### PR TITLE
Upgrade to TypeScript 2.8.1

### PR DESCRIPTION
**Pull request by christiango**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ npm run change`

#### Description of changes
Upgrades fabric react to TypeScript 2.8.1. The new compiler comes with updated typechecks and a new APIs definitions for DOM APIs. The fixes here were made with correctness and safety in mind, so I did not fix any bugs that the new errors introduced.

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx [hunk 1]
```diff
@@ -27,7 +27,7 @@ export class Checkbox extends BaseComponent<ICheckboxProps, ICheckboxState> impl
     boxSide: 'start'
   };
 
-  private _checkBox = createRef<HTMLInputElement>();
+  private _checkBox = createRef<HTMLButtonElement>();
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: Hmm. I think this is correct, but should we just make this more general? `HTMLElement` or something?

File: packages/office-ui-fabric-react/src/components/TextField/TextField.tsx [hunk 2]
```diff
@@ -249,14 +249,14 @@ export class TextField extends BaseComponent<ITextFieldProps, ITextFieldState> i
   /**
    * Gets the selection start of the text field
    */
-  public get selectionStart(): number {
+  public get selectionStart(): number | null {
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/TextField/TextField.tsx]: probably also scary

File: packages/office-ui-fabric-react/src/components/Autofill/Autofill.tsx [hunk 3]
```diff
@@ -33,7 +33,7 @@ export class Autofill extends BaseComponent<IAutofillProps, IAutofillState> impl
     };
   }
 
-  public get cursorLocation(): number {
+  public get cursorLocation(): number | null {
```
[the remark this answers is not in the document]
Comment (christiango) [reply packages/office-ui-fabric-react/src/components/Autofill/Autofill.tsx]: Partners can quickly fix it with a ! operator. The 0 change would be a bigger problem IMO

File: packages/office-ui-fabric-react/src/components/TextField/TextField.tsx (hunk 2 above)
Comment (christiango) [reply packages/office-ui-fabric-react/src/components/TextField/TextField.tsx]: So it already can return null, the typings were just incorrect before. This doesn't change existing behavior, but it does inform partners using strict null checks that they need to do a null check here

File: packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx (hunk 1 above)
Comment (christiango) [reply packages/office-ui-fabric-react/src/components/Checkbox/Checkbox.base.tsx]: Fixed

**Comment by christiango** [comment]
@dzearing  do we need to make any more changes here?

**Comment by christiango** [comment]
Ping @dzearing?

**Comment by manishgarg1** [comment]
Can this affect downstream dependencies in any way? e.g. sp-client?

**Comment by christiango** [comment]
It shouldn't impact downstream partners

**Comment by dzearing** [comment]
merged master to fix the build here.

**Pull request merged**